gitlab/resource_gitlab_repository_file.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabRepositoryFile() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabRepositoryFileCreate,
+		Read:   resourceGitlabRepositoryFileRead,
+		Update: resourceGitlabRepositoryFileUpdate,
+		Delete: resourceGitlabRepositoryFileDelete,
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"file": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"content": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"branch": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"author_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"author_email": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"commit_message": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabRepositoryFileCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	file := d.Get("file").(string)
+
+	options := &gitlab.CreateFileOptions{
+		Branch:        gitlab.String(d.Get("branch").(string)),
+		AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+		AuthorName:    gitlab.String(d.Get("author_name").(string)),
+		Content:       gitlab.String(d.Get("content").(string)),
+		CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+		Encoding:      gitlab.String("base64"),
+	}
+
+	repositoryFile, _, err := client.RepositoryFiles.CreateFile(project, file, options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(repositoryFile.FilePath)
+
+	return resourceGitlabRepositoryFileRead(d, meta)
+}
+
+func resourceGitlabRepositoryFileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	fileID := d.Id()
+	options := &gitlab.GetFileOptions{
+		Ref: gitlab.String(d.Get("branch").(string)),
+	}
+
+	repositoryFile, _, err := client.RepositoryFiles.GetFile(project, fileID, options)
+	if err != nil {
+		return err
+	}
+
+	d.Set("project", project)
+	d.Set("file", repositoryFile.FileName)
+	d.Set("content", repositoryFile.Content)
+	d.Set("branch", repositoryFile.Ref)
+
+	return nil
+}
+
+func resourceGitlabRepositoryFileUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	file := d.Get("file").(string)
+	options := &gitlab.UpdateFileOptions{
+		Branch:        gitlab.String(d.Get("branch").(string)),
+		AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+		AuthorName:    gitlab.String(d.Get("author_name").(string)),
+		Content:       gitlab.String(d.Get("content").(string)),
+		CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+		Encoding:      gitlab.String("base64"),
+		//TODO: add LastCommitID
+	}
+
+	if d.HasChange("branch") {
```

Comment:
I think you'll always want to add these attributes, even if they haven't changed. `branch` in particular because it's required, but the others as well. Maybe for the others you can use `GetOk` instead of `HasChange`?